libr/util/file.c

```diff
@@ -865,44 +902,63 @@ R_API char *r_file_temp (const char *prefix) {
 }
 
 R_API int r_file_mkstemp(const char *prefix, char **oname) {
-	int h;
+	int h = -1;
 	char *path = r_file_tmpdir ();
-	char name[1024] = {0};
 #if __WINDOWS__
-	h = -1;
-	if (GetTempFileNameA (path, prefix, 0, name)) {
-		h = r_sandbox_open (name, O_RDWR|O_EXCL|O_BINARY, 0644);
+	TCHAR name[MAX_PATH + 1];
+	LPTSTR path_ = r_sys_conv_char_to_w32 (path);
+	LPTSTR prefix_ = r_sys_conv_char_to_w32 (prefix);
+
+	if (GetTempFileName (path_, prefix_, 0, name)) {
+		char *name_ = r_sys_conv_w32_to_char (name);
+		h = r_sandbox_open (name_, O_RDWR|O_EXCL|O_BINARY, 0644);
+		if (oname) {
+			if (h != -1) {
+				*oname = name_;
+			} else {
+				*oname = NULL;
+				free (name_);
+			}
+		} else {
+			free (name_);
+		}	
 	}
+	free (path_);
+	free (prefix_);
 #else
+	char name[1024];
+
 	snprintf (name, sizeof (name) - 1, "%s/r2.%s.XXXXXX", path, prefix);
 	mode_t mask = umask (S_IWGRP | S_IWOTH);
 	h = mkstemp (name);
 	umask (mask);
-#endif
 	if (oname) {
 		*oname = (h!=-1)? strdup (name): NULL;
 	}
+#endif
 	free (path);
 	return h;
 }
 
 R_API char *r_file_tmpdir() {
 #if __WINDOWS__
-	char tmpdir[MAX_PATH];
+	TCHAR tmpdir[MAX_PATH + 1];
 	char *path = NULL;
+	DWORD len = 0;
 
-	if (GetTempPathA (sizeof (tmpdir), tmpdir) == 0) {
+	if ((len = GetTempPath (MAX_PATH + 1, tmpdir)) == 0) {
 		path = r_sys_getenv ("TEMP");
 		if (!path) {
 			path = strdup ("C:\\WINDOWS\\Temp\\");
 		}
 	} else {
-		DWORD (WINAPI *glpn)(LPCSTR, LPCSTR, DWORD) = r_lib_dl_sym (GetModuleHandle (TEXT ("kernel32.dll")), "GetLongPathNameA");
+		tmpdir[len] = 0;
```

Comment:
please don't use VLA - they allocate on stack. Create buffer in heap instead please.